README
POLYGLOT(6)                                                        POLYGLOT(6)



NAME
       PolyGlot -  Winboard protocol to UCI protocol adapter
                -  book engine for Polyglot books
                -  a collection of utilities for creating and analyzing
       opening books
                -  a utility for analyzing epd files
                -  a perft counter

SYNOPSIS
       polyglot [configfile] [-noini] [-ec engine] [-ed enginedirectory] [-en
       enginename] [-log true/false] [-lf logfile] [-pg <name>=<value>]* [-uci
       <name>=<value>]*

       polyglot make-book [-pgn inputfile] [-bin outputfile] [-max-ply ply]
       [-min-game games] [-min-score score] [-only-white] [-only-black]
       [-uniform]

       polyglot merge-book -in1 inputfile1 -in2 inputfile2 [-out outputfile]

       polyglot info-book [-bin inputfile] [-exact]

       polyglot dump-book [-bin inputfile] -color color [-out outputfile]

       polyglot [configfile] epd-test [engineoptions] [-epd inputfile]
       [-min-depth depth] [-max-depth depth] [-max-time time] [-depth-delta
       delta]

       polyglot perft [-fen fen] [-max-depth depth]

DESCRIPTION
   PolyGlot as adapter and book engine
       PolyGlot is a "UCI adapter".  It connects a GUI interface (such as
       XBoard, Winboard, Arena or Chessbase) to a UCI chess engine.

       By specifying an opening book (in PolyGlot book format) chess engines
       can transparently use such books.

       PolyGlot understands the two main GUI protocols: UCI and xboard.
       Normally the protocol will be auto detected but this can be overridden
       in the configuration file.

       In xboard mode PolyGlot fully translates between the xboard and UCI
       protocols.  In addition it tries to solve known problems with other
       adapters.  For instance, it detects and reports draws by fifty-move
       rule, repetition, etc ... It also supports Chess960.

       When in UCI mode PolyGlot mostly passes commands from the GUI to the
       engine and vice versa, except that it will play book moves on behalf of
       the engine when the occasion arises.

       The engine options are exported as UCI options in UCI mode and as
       "feature option=" commands in xboard mode. The latter form an extension
       of the xboard protocol as defined by H.G. Muller.

       Options which normally appear in the [PolyGlot] section of the config
       file (see below) are exported as options with their name prefixed by
       "Polyglot". This makes it easy to filter them in the GUI.

       NOTE: Not all options are exported, only those that make sense in the
       given mode.

   Book making utilities
       PolyGlot supports the "PolyGlot opening book format". This is the
       defacto standard non-proprietary opening book format. It is fully
       documented here

       http://alpha.uhasselt.be/Research/Algebra/Toga/book_format.html

       Roughly speaking a PolyGlot opening book is a collection of triples
       (position, move, weight). A "position" is represented by a 64-bit
       Zobrist hash key. The weight is proportional to the probability the
       move should be played.

       Other opening book formats such as ChessBase's .ctg format and Arena's
       .abk format are undocumented and proprietary. They can only be used by
       their own GUIs.

       PolyGlot can compile a pgn file into a binary PolyGlot book and
       furthermore it can merge two such binary books into a third one.

       PolyGlot can also extract some useful information from PolyGlot books.
       The utility "dump-book" dumps the "lines" in a book for a given color.
       By definition a line is a sequence of moves (from the starting
       position) in which the given color makes only book moves and the other
       color makes arbitrary moves (i.e. not necessarily book moves).

       Since a PolyGlot book is built up from positions and not lines there
       may be (and there usually are) many positions in the book that are not
       on a "line" as defined in the previous paragraph. It is convenient to
       call such positions "isolated" positions. The utility "info-book"
       counts such isolated positions.

       Some of the isolated positions are provably unreachable and they could
       in principle be deleted from the book. For example if a book contains
       only the move "e4" in the starting position but also the position after
       "d4 d5" then this last position is provably unreachable since it
       requires white to make a non-book move when a book move is available.
       Such situations arise frequently from the priority rules in merging
       books.

       Unfortunately not all isolated positions are provably unreachable and
       it is difficult to identify the latter. If invoked with "-exact" the
       utility info-book will attempt to count the isolated positions which
       require a player to make a non-book move when a book move is available.
       Due to the possibility of transpositions this is not a fool proof
       method.

   Epd test mode
       In epd test mode, PolyGlot will search positions in an epd file and
       record the number of times the right best move was found.  The
       arguments specify when to stop the search in any given position.

   Perft counts
       A perft count is the number of legal move sequence in a given position
       up to a given depth. PolyGlot can perform such perft counts. It is
       however much slower than other more dedicated programs.

OPTIONS
       When PolyGlot is invoked as an adapter of in epd-test mode it gets its
       options from a config file and then from the command line.  The default
       config file is "polyglot.ini" but an alternative one may be optionally
       included as first argument. The config file format is described below.

       The following engine options may be specified on the command line.

       -noini
           Do not use a config file, even if one was specified on the command
           line.

       -pg The argument is a string of the form <name>=<value>. This option
           will set the Polyglot option <name> to <value>.

       -uci
           The argument is a string of the form <name>=<value>. This option
           will set the engine option <name> to <value>.

       -ec This is an alias for -pg "EngineCommand=<value>"

       -ed This is an alias for -pg "EngineDir=<value>"

       -en This is an alias for -pg "EngineName=<value>"

       -log (default: false)
           This is an alias for -pg "Log=<value>"

       -lf (default: "polyglot.log")
           This is an alias for -pg "LogFile=<value>".

       -wb (default: "true")
           This is an alias for -pg "OnlyWbOptions=<value>".

       When invoked as

   polyglot make-book
       PolyGlot supports the following options

       -pgn (default: "book.pgn")
           Input file in pgn format.

       -bin (default: "book.bin")
           Output file in PolyGlot format.

       -max-ply (default: 1024)
           Specifies the maximum ply-depth of lines included in the book.

       -min-game (default: 3)
           Specifies the minimum number of games that have to contain this
           move for it to be included in the book.

       -min-score (default: 0.0)
           Specifies the minimum score (or weight) this move should have
           received for it to  be included in the book. The score is
           2*(wins)+(draws), globally scaled to fit into 16 bits.

       -only-white
           Include only moves for white in the book.

       -only-black
           Include only moves for black in the book.

       -uniform
           Set all weights to 1. In other words, all moves will be selected
           with equal probability.

       When invoked as

   polyglot merge-book
       PolyGlot supports the following options

       -in1
           First input file (in PolyGlot book format).

       -in2
           Second input file (in PolyGlot book format).

       -out (default: out.bin)
           Output file (in PolyGlot book format).

       Input files are not symmetrical, "in1" has priority over "in2". In
       other words when a position occurs both in "in1" and "in2" only the
       moves and weights from "in1" will be retained in "out".

       When invoked as

   polyglot dump-book
       PolyGlot supports the following options

       -bin (default: book.bin)
           Input file in PolyGlot book format.

       -color
           The color for whom to generate the lines.

       -out (default: book_<color>.txt)
           The name of the output file.

       When invoked as

   polyglot info-book
       PolyGlot supports the following options

       -bin (default: book.bin)
           Input file in PolyGlot book format.

       -exact
           Attempt to count the provably unreachable positions among the
           isolated ones.  Note that this takes a very long time.

       When invoked as

   polyglot epd-test
       (possibly with a config file as first argument) PolyGlot supports
       besides the generic options described above the following additional
       options.

       -max-depth (default: 63)
           Unconditionally stop the search when this depth has been reached.

       -max-time (default: 5.0)
           Unconditionally stop the seach after this amount of time.

       -depth-delta (default: 3)
           Stop the search if the solution as been found and the best move has
           been constant for this many depths, on condition that the mininal
           depth and minimal time have been reached.

       -min-depth (default: 8)
           Minimal search depth when the search is stopped using
           "-depth-delta".

       -min-time (default: 1.0)
           Minimal search time when the search is stopped using
           "-depth-delta".

       When invoked as

   polyglot perft
       PolyGlot supports the following options

       -fen (default: starting position)
           Fen at which to start searching.

       -max-depth (default: 1)
           Maximum depth to search.

CONFIG FILE FORMAT
       There should be a different config file for each engine.

       The config file is in the traditional INI format.

           [PolyGLot]
           option = value
           ...
           [Engine]
           option = value
           ...

       The characters "#" and ";" serve as comment characters.

       Initial and final white space is stripped from option names and values.
       If you need to use characters which have a special meaning to PolyGlot
       (these are ';#[]=\') you should quote them by preceding them with '\'.
       "Quoting" other characters in this way has no effect. In particular the
       use of '\' as a path separator in windows should normally not affected.

   [PolyGlot] section
       This section is used by PolyGlot only.  The engine is unaware of these
       options.  The list of available options is detailed below.

       EngineName (default: UCI name)
           This is the name that will appear in the GUI.  It is cosmetic only.
           You can use different names for tweaked versions of the same
           engine.

       EngineDir (default: ".")
           Full path of the directory where the engine is installed.  You can
           use "." (without the quotes) if you know that PolyGlot will be
           launched in the engine directory or the engine is in the "path" and
           does not need any data file.

       EngineCommand
           Put here the name of the engine executable file.  You can also add
           command-line arguments.  Path searching is used and the current
           directory will be "EngineDir". On Linux the EngineCommand is passed
           to wordexp so that shell quoting rules and expansions are applied.
           On Windows the EngineCommand is simply passed to CreateProcess
           which does its own shell like processing.

       SettingsDir (default: $HOME/.polyglot on Linux; ".\_PG" on Windows)
           The directory where ini files are stored for engines that are
           started with -noini. Such ini files may be created by pushing the
           "Save" button in the Engine settings dialog in WB/XB 4.4.0 and
           higher. As a special exception (for WB/XB 4.4.0 compatibility) this
           directory is also used in case PolyGlot is started with config
           files named "polyglot_1st.ini" or "polyglot_2nd.ini".

       Log (default: false)
           Whether PolyGlot should log all transactions with the interface and
           the engine.  This should be necessary only to locate problems.

       LogFile (default: polyglot.log)
           The name of the log file.  Note that it is put where PolyGlot was
           launched from, not into the engine directory.

           WARNING: Log files are not cleared between sessions, and can become
           very large.  It is safe to remove them though.

       Resign (default: false)
           Set this to "true" if you want PolyGlot to resign on behalf of the
           engine.

           NOTE: Some engines display buggy scores from time to time although
           the best move is correct.  Use this option only if you know what
           you are doing (e.g. you always check the final position of games).

       ResignMoves (default: 3)
           Number of consecutive moves with "resign" score (see below) before
           PolyGlot resigns for the engine.  Positions with only one legal
           move are ignored.

       ResignScore (default: 600)
           This is the score in centipawns that will trigger resign
           "counting".

       ShowPonder (default: true)
           Show search information during engine pondering.  Turning this off
           might be better for interactive use in some interfaces.

       ScoreWhite (default: true)
           Report score from white's point of view in xboard mode.

       KibitzMove (default: false)
           Whether to kibitz when playing a move.

       KibitzPV (default: false)
           Whether to kibitz when the PV is changed (new iteration or new best
           move).

       KibitzCommand (default: "tellall")
           xboard command to use for kibitzing, normally "tellall" for
           kibitzing or "tellothers" for whispering.

       KibitzDelay (default: 5)
           How many seconds to wait before starting kibitzing.  This has an
           effect only if "KibitzPV" is selected, move kibitzes are always
           sent regardless of the delay.

       KibitzInterval (default: 0)
           This is another form of throttling. PolyGlot will usually wait this
           many seconds before doing the next kibitz.

       UCI (default: false)
           If true PolyGlot will not understand xboard commands.

       MateScore (default: 10000)
           Mate score reported to GUI when in xboard mode.

       Book (default: false)
           Indicates whether a PolyGlot book should be used.  This has no
           effect on the engine own book (which can be controlled with the UCI
           option "OwnBook" in the [Engine] section).  In particular, it is
           possible to use both a PolyGlot book and an engine book.  In that
           case, the engine book will be used whenever PolyGlot is out of
           book.  Remember that PolyGlot is unaware of whether the engine is
           itself using a book or not.

       BookFile (default: book.bin)
           The name of the (binary) book file.  Note that PolyGlot will look
           for it in the directory it was launched from, not in the engine
           directory.  Of course, full path can be used in which case the
           current directory does not matter.

       BookRandom (default: true)
           Select moves according to their weights in the book. If false the
           move with the highest weight is selected.

       BookLearn (default: false)
           This is a noop.

       BookDepth (default: 256)
           Stop using the book after this number of moves.

       BookTreshold (default: 5)
           Do not play moves with a weight (probability) lower than this (in
           per mil).

       UseNice (default: false)
           Run the engine at nice level 5, or "NiceValue" if it set.  On some
           operating systems it may be necessary to run the engine at lower
           priority for it to be responsive to commands from PolyGlot while
           searching.

       NiceValue (default: 5)
           Nice levels go from -20 to 20 with 20 being the lowest priority.
           On Unix only root can set negative nice levels. On Windows the
           standard Win32 priority levels are mapped in a sensible way to Unix
           nice levels.

       Affinity (default: -1)
           This a bit vector in which each bit represents the processors that
           a process is allowed to run on. This option works only on Windows.

       STFudge (default: 20)
           PolyGlot will translate "st x" as "go movetime 1000*x-STFudge".
           The rationale is that in the UCI specification the argument of
           movetime is defined as the exact search time whereas the argument
           of the st command is only an upperbound.

       OnlyWbOptions (default: true)
           If true then PolyGlot restricts the options it sends to those that
           are potentially useful for WinBoard.

   Work arounds
       Work arounds are identical to options except that they should be used
       only when necessary.  Their purpose is to try to hide problems with
       various software (not just engines).

       IMPORTANT: Any of these work arounds might be removed in future
       versions of PolyGlot.  You are strongly recommended to contact the
       author of faulty software and truly fix the problem.

       PolyGlot supports the following work arounds:

       UCIVersion (default: 2)
           The default value of 2 corresponds to UCI+.  Use 1 to select plain
           UCI for engines that have problems with UCI+.

       CanPonder (default: false)
           PolyGlot now conforms to the documented UCI behaviour: the engine
           will be allowed to ponder only if it (the engine) declares the
           "Ponder" UCI option.  However some engines which can actually
           ponder do not declare the option.  This work around lets PolyGlot
           know that they can ponder.

       SyncStop (default: false)
           When a ponder miss occurs, Polyglot interrupts the engine and
           IMMEDIATELY launches a new search.  While there should be no
           problem with this, some engines seem confused and corrupt their
           search board.  "SyncStop" forces PolyGlot to wait for the (now
           useless) ponder search to finish before launching the new search.

       PromoteWorkAround (default: false)
           Some engines do not specify a promotion piece, e.g. they send
           "e7e8" instead of the correct "e7e8q".  This work around enables
           the incorrect form (and of course promotes into a queen).

       RepeatPV (default: true)
           When true, PolyGlot repeats the last pv string (which also contains
           score,depth and time usage) it got from the engine. Some engines
           however do not send a new pv string just before sending the move.
           In that case the output of PolyGlot would be inconsistent.  When
           RepeatPV is false PolyGlot does not repeat the last pv string.  Due
           to the way kibitzing is implemented, KibitzMove is disabled in that
           case.

   [Engine] section
       This section contains engine UCI options.  PolyGlot does not understand
       them, but sends the information to the engine at startup (converted to
       UCI form).  You can add any UCI option that makes sense to the engine
       (not just the common options about hash-table size and tablebases).


ChangeLog
========1.4.70b=================	
- merge-book did not work correctly on Windows.
========1.4.69b=================	
- Adapt to latest version of header format.
========1.4.68b=================	
- Make Polyglot merge-book and make-book aware of the new header format.
========1.4.67b=================
- Disable "book learning". Polyglot saves learning information but does not use it itself, probably for a reason. I believe the booklearning concept of Polyglot is fundamentally broken. Learning information is engine specific so it should not be saved in the opening book (it might be saved elsewhere). There are other issues as well, such as copyright and the possibility of the proliferation of polluted books. Opening books can be improved with information from games by using the merge command. This is a form of off line learning which is probably much more effective since one has strict control over what kind of information is added to the book.
========1.4.66b=================
- Fix a printf 32 vs 64 bit format (crash) bug which manifested itself on windows.
========1.4.65b=================
- Bugfix: dump-book with -color black generated the file book_white.txt.
========1.4.64b=================
- Reduce movetime in the implementation of the fixed time per move command, to to account for differences in the wording of the protocols. 
========1.4.63b=================
- Give ponder move as hint.
- Support for egtpath gaviota (this is a bit hacky right now).
========1.4.62b=================
- StringSize was not everywhere the same in Polyglot. This could lead to a buffer overflow in case of very long PV's. 
========1.4.61b=================
- uci_isready replaced by uci_isready_sync. Polyglot should not send commands to the engine while it is syncing.
========1.4.60b=================
- The result string after an illegal move now shows the actual attempted move.
========1.4.59b=================
- Bugfix: the changes in 1.4.57b created a regression where the input buffer of polyglot could overflow with engines producing lots of output very fast. 
========1.4.58b=================
- Implementation of BookDepth in UCI mode.
- Small corrections to the manpage.
========1.4.57b=================
- First attempt at killing engines that do not react to "quit". 
========1.4.56b=================
- Better handling of non-existing engine command on Linux. No changes on Windows.
=========1.4.55b================
- Suppression of some error dialogs.
- Small delay in between final error message and actually quitting.
=========1.4.54b================
- More graceful handling of engine crashes.
- New EngineCommand is now passed through wordexp on Linux.
=========1.4.53b================
- Multipv code by HGM. 
=========1.4.52b================
- Reexport SyncStop. It is apparently necessary for the Aquarium Book Adapter.
=========1.4.51b================
- Support for new nps command. Code provided by Harm Geert Muller. 
=========1.4.50b================
- Change in architecture. Persistence is disabled. The user is required to push the "Save" button explicitly to save the settings. What is retained is that in case PG is invoked with -noini (or as a special exception for WB 4.4.0 when the config files are polyglot_1st.ini or polyglot_2nd.ini) then PG uses a default config file whose name is derived from the engine name. 
- New option: OnlyWbOptions (default: true). Restrict the options that PG sends to those that are potentially useful for WinBoard.
=========1.4.47b================
- Try to detect if the engine is not an UCI engine.
=========1.4.46b================
- The default persistence directory on Windows is now _PG. The names of the engine specific ini files are mangled so as not to contain spaces. 
- PG now tries to guess if a string option is really a file or a path (like in the Arena GUI). 
=========1.4.45b================
- The engine specific config files (<myname>.ini) are now stored in a separate directory. $HOME/.polyglot on Linux and "./Polyglot Settings" on Windows. 
- The engine specific config files are now complete config files. So they can be used in a stand alone fashion.
- Implemented a minor hack to make xboard show error messages during startup. 
=========1.4.44b================
- "Implementation" of new buttons and controls.
- The "Persist" option is now persistent. That is if you turn of "Persist" then as a special case PG will remember this.
- There is now a "reset" button which restores the default options. The code is commented out since WB does not support it yet.
=========1.4.42b================
- SaveSettingsOnExit/SaveFile renamed as Persist/PersistFile.
- Many bug fixes in the persistence feature.
- The PG_<myname>.ini file now contains only options that have been changed. So the resulting ini files are much smaller. This makes this version of PG usable on WB 4.3.15 as now the engine specific ini file will be empty (the user has no way to change options).
=========1.4.41b================
- Some segfaults fixed in case of corrupt config files.
- Persistence of options. If the option SaveSettingsOnExit is true (the default) PolyGlot now saves its options in a file whose name is given by the option SaveFile which by default is equal to the engine name, prefixex by PG_ and suffixed by ".ini".
=========1.4.39b================
- Refactoring of config file parsing
- Fix crash bug in epd-test
- Do not export Chess960. This is handled completely internally. 
- Do not export UCI_Opponent and UCI_Chess960. These are handled internally.
- PG can now also be comfortably configured through command line options (i.e. without using a config file). 
- Allow quotes in configfile (as in Fonzy Bleumer's PG). 
=========1.4.37b================
- Two new options: BookDepth and BookTreshold. BookDepth limits the number of moves the book is used and BookTreshold sets the minimum weight a book move should have.
=========1.4.36b================
- Bugfix: option = 0/1 was translated incorrectly.

epd.c

// epd.c

// includes

#include <errno.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "board.h"
#include "engine.h"
#include "epd.h"
#include "fen.h"
#include "line.h"
#include "main.h"
#include "move.h"
#include "move_legal.h"
#include "option.h"
#include "parse.h"
#include "san.h"
#include "uci.h"
#include "util.h"


// macros

#define  StringSize 4096

// constants

static const bool UseDebug = FALSE;
static const bool UseTrace = FALSE;

// variables

static int MinDepth;
static int MaxDepth;

static double MaxTime;
static double MinTime;

static int DepthDelta;

static int FirstMove;
static int FirstDepth;
static int FirstSelDepth;
static int FirstScore;
static double FirstTime;
static uint64 FirstNodeNb;
static move_t FirstPV[LineSize];

static int LastMove;
static int LastDepth;
static int LastSelDepth;
static int LastScore;
static double LastTime;
static uint64 LastNodeNb;
static move_t LastPV[LineSize];

static my_timer_t Timer[1];

// prototypes

static void epd_test_file  (const char file_name[]);

static bool is_solution    (int move, const board_t * board, const char bm[], const char am[]);
static bool string_contain (const char string[], const char substring[]);

static bool engine_step    ();

// functions

// epd_test()

void epd_test(int argc, char * argv[]) {

   int i;
   const char * epd_file;

   epd_file = NULL;
   my_string_set(&epd_file,"wac.epd");

   MinDepth = 8;
   MaxDepth = 63;

   MinTime = 1.0;
   MaxTime = 5.0;

   DepthDelta = 3;

   for (i = 1; i < argc; i++) {

      if (FALSE) {

      } else if (my_string_equal(argv[i],"epd-test")) {

         // skip

      } else if (my_string_equal(argv[i],"-epd")) {

         i++;
         if (argv[i] == NULL) my_fatal("epd_test(): missing argument\n");

         my_string_set(&epd_file,argv[i]);

      } else if (my_string_equal(argv[i],"-min-depth")) {

         i++;
         if (argv[i] == NULL) my_fatal("epd_test(): missing argument\n");

         MinDepth = atoi(argv[i]);

      } else if (my_string_equal(argv[i],"-max-depth")) {

         i++;
         if (argv[i] == NULL) my_fatal("epd_test(): missing argument\n");

         MaxDepth = atoi(argv[i]);

      } else if (my_string_equal(argv[i],"-min-time")) {

         i++;
         if (argv[i] == NULL) my_fatal("epd_test(): missing argument\n");

         MinTime = atof(argv[i]);

      } else if (my_string_equal(argv[i],"-max-time")) {

         i++;
         if (argv[i] == NULL) my_fatal("epd_test(): missing argument\n");

         MaxTime = atof(argv[i]);

      } else if (my_string_equal(argv[i],"-depth-delta")) {

         i++;
         if (argv[i] == NULL) my_fatal("epd_test(): missing argument\n");

         DepthDelta = atoi(argv[i]);

      } else {

         my_fatal("epd_test(): unknown option \"%s\"\n",argv[i]);
      }
   }

   if(MinTime>MaxTime){
       MaxTime=MinTime;
   }
  
   epd_test_file(epd_file);
}

// epd_test_file()

static void epd_test_file(const char file_name[]) {

   FILE * file;
   int hit, tot;
   char epd[StringSize];
   char am[StringSize], bm[StringSize], id[StringSize];
   board_t board[1];
   char string[StringSize];
   int move;
   char pv_string[StringSize];
   bool correct;
   double depth_tot, time_tot, node_tot;
   int line=0;

   ASSERT(file_name!=NULL);

   // init

   file = fopen(file_name,"r");
   if (file == NULL) my_fatal("epd_test_file(): can't open file \"%s\": %s\n",file_name,strerror(errno));

   hit = 0;
   tot = 0;

   depth_tot = 0.0;
   time_tot = 0.0;
   node_tot = 0.0;

   printf("\nEngineName=%s\n",option_get_string(Option,"EngineName"));

   printf("\n[Search parameters: MaxDepth=%d   MaxTime=%.1f   DepthDelta=%d   MinDepth=%d   MinTime=%.1f]\n\n",MaxDepth,MaxTime,DepthDelta,MinDepth,MinTime);

   // loop

   while (my_file_read_line(file,epd,StringSize)) {
       line++;
       if(my_string_whitespace(epd)) continue;
      if (UseTrace) printf("%s\n",epd);

      if (!epd_get_op(epd,"am",am,StringSize)) strcpy(am,"");
      if (!epd_get_op(epd,"bm",bm,StringSize)) strcpy(bm,"");
      if (!epd_get_op(epd,"id",id,StringSize)) strcpy(id,"");

      if (my_string_empty(am) && my_string_empty(bm)) {
          my_fatal("epd_test(): no am or bm field at line %d\n",line);
      }

      // init

      uci_send_ucinewgame(Uci);
      uci_send_isready_sync(Uci);

      ASSERT(!Uci->searching);

      // position
      if (!board_from_fen(board,epd)) ASSERT(FALSE);
      if (!board_to_fen(board,string,StringSize)) ASSERT(FALSE);

      engine_send(Engine,"position fen %s",string);

      // search

      my_timer_start(Timer); // also resets
      
      // which ones of the next two alternatives is best?
      engine_send(Engine,"go movetime %.0f depth %d",MaxTime*1000.0,MaxDepth);
      //engine_send(Engine,"go infinite");

      // engine data

      board_copy(Uci->board,board);

      uci_clear(Uci);
      Uci->searching = TRUE;
      Uci->pending_nb++;

      FirstMove = MoveNone;
      FirstDepth = 0;
      FirstSelDepth = 0;
      FirstScore = 0;
      FirstTime = 0.0;
      FirstNodeNb = 0;
      line_clear(FirstPV);

      LastMove = MoveNone;
      LastDepth = 0;
      LastSelDepth = 0;
      LastScore = 0;
      LastTime = 0.0;
      LastNodeNb = 0;
      line_clear(LastPV);

      // parse engine output

      while (!engine_eof(Engine) && engine_step()) {
          bool stop=FALSE;

         // stop search?
//          printf("Uci->time=%.2f time=%.2f\n",Uci->time,my_timer_elapsed_real(Timer));
          if (Uci->depth > MaxDepth){
              my_log("POLYGLOT Maximum depth %d reached\n",MaxDepth);
              stop=TRUE;
          }else if(my_timer_elapsed_real(Timer) >= MaxTime){
              my_log("POLYGLOT Maximum search time %.2fs reached\n",MaxTime);
              stop=TRUE;
          }else if(Uci->depth - FirstDepth >= DepthDelta){
              if(Uci->depth > MinDepth){
                  if(Uci->time >= MinTime){
                      if(is_solution(FirstMove,board,bm,am)){
                          my_log("POLYGLOT Solution found\n",MaxTime);
                          stop=TRUE;
                      }
                  }
              }
          }
          if(stop){
              my_log("POLYGLOT Stopping engine\n");
              engine_send(Engine,"stop");
              break;
          }
      }
      
      move = FirstMove;
      correct = is_solution(move,board,bm,am);

      if (correct) hit++;
      tot++;

      if (correct) {
          depth_tot += ((double)FirstDepth);
         time_tot += FirstTime;
         node_tot += ((double)((sint64)FirstNodeNb));
      }

      printf("%2d: %-15s %s %4d",tot,id,correct?"OK":"--",hit);

      if (!line_to_san(LastPV,Uci->board,pv_string,StringSize)) ASSERT(FALSE);
      printf(" score=%+6.2f    pv [D=%2d, T=%7.2fs, N=%6dk] =%s\n",((double)LastScore)/100.0,FirstDepth,FirstTime,(int)FirstNodeNb/1000,pv_string);
   }

   printf("\nscore=%d/%d",hit,tot);

   if (hit != 0) {

       depth_tot /= ((double)hit);
       time_tot /= ((double)hit);
       node_tot /= ((double)hit);

      printf(" [averages on correct positions: depth=%.1f time=%.2f nodes=%.0f]",depth_tot,time_tot,node_tot);
   }

   printf("\n");

   fclose(file);
   quit();
}

// is_solution()

static bool is_solution(int move, const board_t * board, const char bm[], const char am[]) {

   char move_string[256];
   bool correct;

   ASSERT(move!=MoveNone);
   ASSERT(bm!=NULL);
   ASSERT(am!=NULL);

   if (!move_is_legal(move,board)) {
      board_disp(board);
      move_disp(move,board);
      printf("\n\n");
   }

   ASSERT(move_is_legal(move,board));

   if (!move_to_san(move,board,move_string,256)) ASSERT(FALSE);

   correct = FALSE;
   if (!my_string_empty(bm)) {
      correct = string_contain(bm,move_string);
   } else if (!my_string_empty(am)) {
      correct = !string_contain(am,move_string);
   } else {
      ASSERT(FALSE);
   }

   return correct;
}

// epd_get_op()

bool epd_get_op(const char record[], const char opcode[], char string[], int size) {

   char op[256];
   int len;
   const char *p_start, *p_end;

   ASSERT(record!=NULL);
   ASSERT(opcode!=NULL);
   ASSERT(string!=NULL);
   ASSERT(size>0);

   // find the opcode

   sprintf(op," %s ",opcode); 

   p_start = strstr(record,op);
   if (p_start == NULL){
          sprintf(op,";%s ",opcode); 
          p_start = strstr(record,op);
          if (p_start == NULL){
              return FALSE;
          }
   }

   // skip the opcode

   p_start += strlen(op);

   // find the end
   p_end = strchr(p_start,';');
   if (p_end == NULL) return FALSE;

   // calculate the length

   len = p_end - p_start;
   if (size < len+1) my_fatal("epd_get_op(): size < len+1\n");

   strncpy(string,p_start,len);
   string[len] = '\0';

   return TRUE;
}

// string_contain()

static bool string_contain(const char string[], const char substring[]) {

   char new_string[StringSize], *p;

   strcpy(new_string,string); // HACK

   for (p = strtok(new_string," "); p != NULL; p = strtok(NULL," ")) {
      if (my_string_equal(p,substring)) return TRUE;
   }

   return FALSE;
}

// engine_step()

static bool engine_step() {

   char string[StringSize];
   int event;

   engine_get(Engine,string);
   event = uci_parse(Uci,string);

   if ((event & EVENT_MOVE) != 0) {

      return FALSE;
   }

   if ((event & EVENT_PV) != 0) {

      LastMove = Uci->best_pv[0];
      LastDepth = Uci->best_depth;
      LastSelDepth = Uci->best_sel_depth;
      LastScore = Uci->best_score;
      LastTime = Uci->time;
      LastNodeNb = Uci->node_nb;
      line_copy(LastPV,Uci->best_pv);

      if (LastMove != FirstMove) {
         FirstMove = LastMove;
         FirstDepth = LastDepth;
         FirstSelDepth = LastSelDepth;
         FirstScore = LastScore;
         FirstTime = LastTime;
         FirstNodeNb = LastNodeNb;
         line_copy(FirstPV,LastPV);
      }
   }

   return TRUE;
}

// end of epd.cpp

